### PR TITLE
Docker Updates

### PR DESCRIPTION
**Pull request by jssmith**
This change helps improve Docker usability:

- Fixes broken paths in Dockerfile
- Use git archive to export Ray sources to container
- `start_ray.sh` does not exit when used within Docker
- Updated "Install on Docker" instructions
- Created "Using Ray on a Docker Cluster" instructions

Note that this change does not include presently linking to Docker instructions from the top-level README. This is something we should do as soon as we have confidence the instructions work smoothly for people.

### CODE REVIEW

**Comment by robertnishihara** [comment]
We should link to these instructions from README.md, right?

**Comment by jssmith** [comment]
@robertnishihara, ah, I was just updating the description. I'd like to see a few people try it before we link to the README.md, but perhaps we can pass that mark in this code review cycle.

File: doc/install-on-docker.md [hunk 1]
```diff
@@ -1,10 +1,53 @@
 # Installation on Docker
 
-You can install Ray on any platform that runs Docker. We do not presently publish Docker images for Ray, but you can build them yourself using the Ray distribution. Using Docker can provide a reliable way to get up and running quickly.
+You can install Ray on any platform that runs Docker. We do not presently publish Docker images for Ray, but you can build them yourself using the Ray distribution.
+
+Using Docker can streamline the build process reliable way to get up and running quickly.
```
Comment (robertnishihara) [inline doc/install-on-docker.md]: "the build process *in* a reliable way"?

File: doc/install-on-docker.md [hunk 2]
```diff
@@ -1,10 +1,53 @@
...
 
 ## Install Docker
 
-The Docker Platform release is available for Mac, Windows, and Linux platforms. Please download the appropriate version from the [Docker website](https://www.docker.com/products/overview#/install_the_platform).
+### Mac, Linux, Windows platforms
+
+The Docker Platform release is available for Mac, Windows, and Linux platforms. Please download the appropriate version from the [Docker website](https://www.docker.com/products/overview#/install_the_platform) and follow the corresponding installation instructions.
```
Comment (robertnishihara) [inline doc/install-on-docker.md]: I remember thinking/hearing that the official docker installation instructions for Ubuntu are bad and that the ones to follow are here https://www.digitalocean.com/community/tutorials/how-to-install-and-use-docker-on-ubuntu-16-04

Have you ever had any trouble with the official docker instructions?

File: doc/install-on-docker.md [hunk 3]
```diff
@@ -1,10 +1,53 @@
...
+
+### Docker installation on EC2 with Ubuntu
```
Comment (robertnishihara) [inline doc/install-on-docker.md]: These instructions are different from the ones on the Docker website. E.g., the official Docker instructions do not do `sudo apt-get -y dist-upgrade`.

Why the difference?

File: doc/install-on-docker.md [hunk 4]
```diff
@@ -23,21 +66,45 @@ cd ray
...
+```
+
 
 ## Launch Ray in Docker
 
 Start out by launching the deployment container.
 
 ```
-docker run --shm-size=1024m -t -i ray-project/ray:deploy
+docker run --shm-size=<shm-size> -t -i ray-project/ray:deploy
```
Comment (robertnishihara) [inline doc/install-on-docker.md]: This command doesn't work, I think you need to remove the `ray:` and just use `ray-project/deploy`

File: doc/install-on-docker.md [hunk 5]
```diff
@@ -23,21 +66,45 @@ cd ray
...
 ```
 
+
 ## Test if the installation succeeded
 
-To test if the installation was successful, try running some tests.
+To test if the installation was successful, try running some tests. Within the container shell enter the following commands:
 
 ```
 python test/runtest.py # This tests basic functionality.
```
Comment (robertnishihara) [inline doc/install-on-docker.md]: I just tried this and it told me 

```
Traceback (most recent call last):
  File "test/runtest.py", line 127, in <module>
    raise Exception("You have an older version of cloudpickle that is not able to serialize namedtuples. Try running \n\n{}\n\n".format(cloudpickle_command))
Exception: You have an older version of cloudpickle that is not able to serialize namedtuples. Try running 

pip install --upgrade cloudpickle
```

We should make sure that we're installing the correct version of cloudpickle in the docker image. For that it should be sufficient to do `pip install cloudpickle` somewhere.

File: doc/using-ray-and-docker-on-a-cluster.md [hunk 6]
```diff
@@ -0,0 +1,229 @@
...
+We recommend that you read the official Docker documentation for [Building your own image](https://docs.docker.com/engine/getstarted/step_four/) ahead of starting this section.
+
+This is an example template Dockerfile for a Ray application:
+
+```
+# Application Dockerfile template
+FROM ray-project/deploy
+RUN git clone https://github.com/my-user/my-project.git
+RUN ./my-project/install.sh
+```
```
Comment (robertnishihara) [inline doc/using-ray-and-docker-on-a-cluster.md]: I don't understand this example. What is `my-project` supposed to be? And where would the installation of Ray happen?

File: doc/using-ray-and-docker-on-a-cluster.md [hunk 7]
```diff
@@ -0,0 +1,229 @@
...
+To start Ray on the head node run the following command:
+
+```
+eval $(aws ecr get-login --region <region>)
+docker run \
+    -d --shm-size=<shm-size> --net=host \
+    <repository-uri> \
+    /ray/scripts/start_ray.sh --head \
+        --redis-port=6379 \
+        --num-workers=<num-workers>
```
Comment (robertnishihara) [inline doc/using-ray-and-docker-on-a-cluster.md]: shall we also have people use a fixed object store port?

Comment (robertnishihara) [inline doc/using-ray-and-docker-on-a-cluster.md]: Both here and on the worker nodes

File: doc/install-on-docker.md [hunk 8]
```diff
@@ -59,7 +126,7 @@ docker run --shm-size=1024m -t -i ray-project/ray:examples
```
Comment (robertnishihara) [inline doc/install-on-docker.md]: the command above

```
docker run --shm-size=1024m -t -i ray-project/ray:examples
```

should be 

```
docker run --shm-size=1024m -t -i ray-project/examples
```

File: docker/base-deps/Dockerfile [hunk 9]
```diff
@@ -10,5 +10,5 @@ RUN echo 'export PATH=/opt/conda/bin:$PATH' > /etc/profile.d/conda.sh \
     && /bin/bash /tmp/anaconda.sh -b -p /opt/conda \
     && rm /tmp/anaconda.sh
 ENV PATH "/opt/conda/bin:$PATH"
-RUN conda install libgcc
+RUN conda install -y libgcc
 RUN pip install --upgrade pip
```
Comment (robertnishihara) [inline docker/base-deps/Dockerfile]: probably want do pip install some stuff (at least cloudpickle) here

**Review by robertnishihara — COMMENTED** [review]
Looks great! I haven't had a chance to try out the cluster version yet.

File: doc/install-on-docker.md [hunk 10]
```diff
@@ -68,7 +135,7 @@ See the [Hyperparameter optimization documentation](../examples/hyperopt/README.
 ### Batch L-BFGS
 
 ```
-cd ~/ray/examples/lbfgs/
+cd /ray/examples/lbfgs/
```
Comment (robertnishihara) [inline doc/install-on-docker.md]: When I run this example I get a lot of errors like the following.

```
W tensorflow/core/platform/cpu_feature_guard.cc:45] The TensorFlow library wasn't compiled to use SSE3 instructions, but these are available on your machine and could speed up CPU computations.
W tensorflow/core/platform/cpu_feature_guard.cc:45] The TensorFlow library wasn't compiled to use SSE4.1 instructions, but these are available on your machine and could speed up CPU computations.
W tensorflow/core/platform/cpu_feature_guard.cc:45] The TensorFlow library wasn't compiled to use SSE4.2 instructions, but these are available on your machine and could speed up CPU computations.
W tensorflow/core/platform/cpu_feature_guard.cc:45] The TensorFlow library wasn't compiled to use AVX instructions, but these are available on your machine and could speed up CPU computations.
W tensorflow/core/platform/cpu_feature_guard.cc:45] The TensorFlow library wasn't compiled to use AVX2 instructions, but these are available on your machine and could speed up CPU computations.
W tensorflow/core/platform/cpu_feature_guard.cc:45] The TensorFlow library wasn't compiled to use FMA instructions, but these are available on your machine and could speed up CPU computations.
[INFO] (/ray/src/photon/photon_scheduler.c:202) Started worker with pid 4463
W tensorflow/core/platform/cpu_feature_guard.cc:45] The TensorFlow library wasn't compiled to use SSE3 instructions, but these are available on your machine and could speed up CPU computations.
```

File: doc/install-on-docker.md (hunk 2 above)
Comment (jssmith) [inline doc/install-on-docker.md]: Yeah, the official Linux instructions are not very relevant to our use case. I'll add a link to the alternate instructions.

File: doc/install-on-docker.md (hunk 3 above)
Comment (jssmith) [inline doc/install-on-docker.md]: The official Linux instructions seem to focus on connecting to Dockerhub and getting a repository set up. These are simple instructions to get Docker installed and to get it to run without sudo.

The `dist-upgrade` command here is likely optional. It' something I always do just so I'm running the latest patched version of the OS. I know things have at times broken without it but I don't remember whether it is necessary right now. I think it's probably safest to leave it in here.

File: doc/install-on-docker.md (hunk 10 above)
Comment (jssmith) [inline doc/install-on-docker.md]: I looked at forums a bit and these warnings seem to come from some recent changes in Tensorflow. I don't believe this is related to Ray or Docker and it something we may want to come back to.

File: docker/base-deps/Dockerfile (hunk 9 above)
Comment (jssmith) [inline docker/base-deps/Dockerfile]: Will add cloudpickle.

File: doc/using-ray-and-docker-on-a-cluster.md (hunk 7 above)
Comment (jssmith) [inline doc/using-ray-and-docker-on-a-cluster.md]: Is there a port number we have standardized on?

Comment (robertnishihara) [inline doc/using-ray-and-docker-on-a-cluster.md]: Not yet. Maybe just choose one? E.g., `7183`.

**Comment by robertnishihara** [comment]
Btw, I ran the docker instructions without the `sudo apt-get -y dist-upgrade` and it worked fine.

**Pull request merged**